source/common/common/token_bucket_impl.cc

```diff
@@ -103,4 +103,13 @@ double AtomicTokenBucketImpl::timeNowInSeconds() const {
   return std::chrono::duration<double>(time_source_.monotonicTime().time_since_epoch()).count();
 }
 
+std::chrono::milliseconds AtomicTokenBucketImpl::nextTokenAvailable() const {
+  // If there are tokens available, return immediately.
+  if (remainingTokens() >= 1) {
+    return std::chrono::milliseconds(0);
+  }
+
+  return std::chrono::milliseconds(static_cast<uint64_t>(std::ceil((1 / fill_rate_) * 1000)));
```

Comment:
This calculates the time based on when the next single token will be available, but doesn't take into the account partial fills or time that has already elapsed in the current interval. I feel that a better approach might be to do something like this:

```cpp
namespace {
  constexpr auto MILLISECONDS_PER_SECOND = 1000;
  constexpr auto MINIMUM_TOKENS_FOR_REQUEST = 1;
  // ...
}
```

And then,

```cpp
...

  // Calculate time since the last fill
  double current_time = timeNowInSeconds();
  double time_since_last_fill = std::fmod(current_time, fill_interval_);
  double time_until_next_fill = fill_interval_ - time_since_last_fill;
  
  return std::chrono::milliseconds(static_cast<uint64_t>(time_until_next_fill * MILLISECONDS_PER_SECOND));
```